pkg/local_object_storage/metabase/version.go

```diff
@@ -25,32 +26,39 @@ var versionKey = []byte("version")
 // the current code version.
 var ErrOutdatedVersion = logicerr.New("invalid version, resynchronization is required")
 
-func (db *DB) checkVersion(tx *bbolt.Tx) error {
-	stored, knownVersion := getVersion(tx)
+func (db *DB) syncVersion() error {
+	var migrateFromV3 bool
+	if err := db.boltDB.Update(func(tx *bbolt.Tx) error {
+		stored, knownVersion := getVersion(tx)
 
-	switch {
-	case !knownVersion:
-		// new database, write version
-		return updateVersion(tx, currentMetaVersion)
-	case stored == currentMetaVersion:
-		return nil
-	case stored > currentMetaVersion:
-		return fmt.Errorf("%w: expected=%d, stored=%d", ErrOutdatedVersion, currentMetaVersion, stored)
-	}
-
-	// Outdated, but can be migrated.
-	for i := stored; i < currentMetaVersion; i++ {
-		migrate, ok := migrateFrom[i]
-		if !ok {
+		switch {
+		case !knownVersion:
+			// new database, write version
+			return updateVersion(tx, currentMetaVersion)
+		case stored == currentMetaVersion:
+			return nil
+		case stored > currentMetaVersion:
+			return fmt.Errorf("%w: expected=%d, stored=%d", ErrOutdatedVersion, currentMetaVersion, stored)
+		case stored < 2:
 			return fmt.Errorf("%w: expected=%d, stored=%d", ErrOutdatedVersion, currentMetaVersion, stored)
 		}
 
-		err := migrate(db, tx)
-		if err != nil {
-			return fmt.Errorf("migrating from meta version %d failed, consider database resync: %w", i, err)
+		// Outdated, but can be migrated.
+		if stored == 2 {
+			if err := migrateFrom2Version(db, tx); err != nil {
+				return fmt.Errorf("migrating from meta version 2 failed, consider database resync: %w", err)
+			}
+		}
+		migrateFromV3 = true
```

Comment:
You better expose stored/current and handle it outside, this code will be wrong wtih v5 immediately, previous one was more generic. Or at least checked the stored immediately, if it's 4 already this is not needed.

The scheme with maps is still applicable, btw. Just split the getter from the changer, `migrateFrom2Version` would just be wrapped in another tx.

Reply:
> this code will be wrong wtih v5 immediately

V4+ is caught above

Reply:
> You better expose stored/current and handle it outside

u mean read version in one tx, then update one-by-one in separate ones? or what

Reply:
Yes, exactly. If we can't have a single transaction in every case we can just change the migration function signature.

Reply:
ok, will do. I just wanted to leave previous behavior untouched as much as possible

Reply:
Looks like we have less lines changed overall now which is good, a wrapper for old migrator is not a problem.